### PR TITLE
feat: collect email for subscription

### PR DESCRIPTION
**Pull request by tejaskh3**
Date: Sept 3, 2024

Developer Name: [Tejas](@tejaskh3 )

## Design doc
[DOC](https://docs.google.com/document/d/1KJ5H7KG_k8u2EXb4EBqo9PzovdBxiDGyzHVgKi1_8f0/edit#heading=h.x5uxodod6vzn)
---

## Issue Ticket Number

Please don't close this issue.
https://github.com/Real-Dev-Squad/website-backend/issues/2123
## Description
This PR includes two APIs
In this PR we have added APIs to collect user email and phone number to make them subscribe to RDS notifications, and even unsubscribe as well. 
and to send email to a user.

### Documentation Updated?

- [ ] Yes
- [x] No

### Under Feature Flag

- [ ] Yes
- [x] No

### Database Changes

- [x] Yes
- [ ] No

### Breaking Changes

- [ ] Yes
- [x] No

### Development Tested?

- [x] Yes
- [ ] No

## Screenshots

<details>
  <summary>Click to view Screenshots</summary>

  Screenshot 1
image
image
image

  Screenshot 3

</details>

## Test Coverage

<details>
  <summary>Click to view Test Coverage</summary>

  - **Router**  
    Router Test Coverage

  - **Validator**  
    Validator Test Coverage
  - **Controller**  
    Controller Test Coverage

</details>

## Additional Notes
Here is the API contract for the feature.
https://github.com/Real-Dev-Squad/website-api-contracts/pull/189

### CODE REVIEW

File: routes/subscription.js
Comment (tejaskh3) [reply routes/subscription.js:9]: we are not using rate limiting right now.

File: routes/subscription.js
Comment (tejaskh3) [reply routes/subscription.js:11]: we are not using rate limiting right now.

File: routes/subscription.js
Comment (tejaskh3) [reply routes/subscription.js:10]: we are not using rate limiting right now.

File: routes/subscription.js
Comment (tejaskh3) [reply routes/subscription.js:1]: Every other file here is in .js, that's why I too chose .js. I'll change it.

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts]: Done

File: routes/subscription.js
Comment (tejaskh3) [reply routes/subscription.js:1]: Changed, please check

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:1]: we are now, please check.

**Comment by prakashchoudhary07** [comment]
Please attacht the design doc for it?

**Comment by tejaskh3** [comment]
> Please attacht the design doc for it?

Attached sir. please check

File: config/custom-environment-variables.js
Comment (iamitprakash) [inline config/custom-environment-variables.js:59]: what is RDS_EMAIL

File: controllers/subscription.ts
Comment (iamitprakash) [inline controllers/subscription.ts:24]: is this same for 4xx

**Review by iamitprakash — CHANGES_REQUESTED** [review]
why this is not under FF

File: routes/subscription.ts
Comment (iamitprakash) [inline routes/subscription.ts:11]: is this route name correct (I am not sure about this)

File: test/integration/subscription.test.js
Comment (iamitprakash) [inline test/integration/subscription.test.js:117]: where is test for 400

File: test/unit/middlewares/subscription-validator.test.js
Comment (iamitprakash) [inline test/unit/middlewares/subscription-validator.test.js:55]: where is test for email is not in correct formate

File: routes/subscription.ts
Comment (tejaskh3) [reply routes/subscription.ts:11]: Yes sir. I have even tested this.

File: config/default.js
Comment (yesyash) [inline config/default.js:32]: - are the steps to generate this written in any issue ticket? do we have to complete it before the pr is merged to develop?

File: config/default.js
Comment (yesyash) [inline config/default.js:29]: - should we name this email instead of user as it is not very indicative of what it represents?

File: config/default.js
Comment (yesyash) [inline config/default.js:30]: - let's not use abbreviations if the full text is not too long.

File: config/development.js
Comment (yesyash) [inline config/development.js:37]: - same as in `config/default.js`.

File: controllers/subscription.ts
Comment (yesyash) [inline controllers/subscription.ts:20]: - are we creating a new resource here?

File: controllers/subscription.ts
Comment (yesyash) [inline controllers/subscription.ts:58]: - is this format approved by ankush?

File: controllers/subscription.ts
Comment (yesyash) [inline controllers/subscription.ts:60]: - what does "after approving this pr" mean? are we going to send a test mail to @iamitprakash before the pr is merged to develop?
- who's email is hardcoded here and is it temporary? if so, when will it be removed, and what would the flow look like after that?

File: routes/subscription.ts
Comment (yesyash) [reply routes/subscription.ts:11]: - @tejaskh3 can you write down the route names in the reply please.

Comment (tejaskh3) [reply routes/subscription.ts:11]: yeah sure @yesyash **subscription/send-email**

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:58]: Not now, that is even not our priority right now, we just want to send hello on Tejas sir email as of now, our focus is to collect email and phone numbers first.

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:60]: 1. "After approving the PR" we would have just update email through another small  PR from @iamitprakash's email.
2. This is my another email for local testing. 

If you and @iamitprakash want I can now put @iamitprakash's email here, in order to save our time for next PR.

File: controllers/subscription.ts
Comment (iamitprakash) [reply controllers/subscription.ts:20]: > yes, we are adding users email and phone number here.

is this what we decided

Comment (tejaskh3) [reply controllers/subscription.ts:20]: Yes sir, we decided to add email and phone number in user collection.

File: test/unit/middlewares/subscription-validator.test.js
Comment (tejaskh3) [reply test/unit/middlewares/subscription-validator.test.js:55]: added sir, 

But I have a question here, if there was no test for "email is not in correct format" earlier, how was the coverage 100%?

**Comment by tejaskh3** [comment]
> why this is not under FF

Done sir.

File: config/default.js
Comment (tejaskh3) [reply config/default.js]: changed.

File: config/development.js
Comment (tejaskh3) [reply config/development.js]: done.

File: config/default.js
Comment (tejaskh3) [reply config/default.js]: I haven't made it yet, but will create it before making these commits to main.

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts]: @yesyash I have Changed the email after discussing with Tejas sir in last call.

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:29]: yes sir

File: config/default.js
Comment (yesyash) [reply config/default.js]: > I haven't made it yet, but will create it before making these commits to main.

where is the ticket for this ?
please make one if not made already and add it in the pr description

File: controllers/subscription.ts
Comment (yesyash) [reply controllers/subscription.ts]: - add a todo here to get this format approved by ankush after the email collection is working.

File: controllers/subscription.ts
Comment (yesyash) [reply controllers/subscription.ts:25]: @tejaskh3 does this not just update the phone number, email, and isSubscribed status in the user collection? if so how are we creating a new resource here?

File: routes/index.ts
Comment (yesyash) [inline routes/index.ts:42]: - should this be `subscription` or `subscriptions`?

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:25]: yes we are creating because they were not present their earlier

File: routes/index.ts
Comment (tejaskh3) [reply routes/index.ts:42]: subscription sir

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts]: https://github.com/Real-Dev-Squad/website-backend/issues/2210 i have created an issue for this sir.

File: controllers/subscription.ts
Comment (tejaskh3) [inline controllers/subscription.ts:69]: this is for test only after @iamitprakash's permission we are adding this here.

File: controllers/subscription.ts
Comment (ankushdharkar) [inline controllers/subscription.ts:10]: Is this safe from a security point of view?

File: controllers/subscription.ts
Comment (ankushdharkar) [inline controllers/subscription.ts:14]: This should move to the top of function

File: controllers/subscription.ts
Comment (ankushdharkar) [inline controllers/subscription.ts:34]: Same comment as above.

File: controllers/subscription.ts
Comment (ankushdharkar) [inline controllers/subscription.ts:9]: Please reduce the coverage on try-catch

File: controllers/subscription.ts
Comment (ankushdharkar) [inline controllers/subscription.ts:19]: How are we handling multiple subscriptions for the same user? Is it allowed?

File: controllers/subscription.ts
Comment (ankushdharkar) [inline controllers/subscription.ts:49]: Should this be in a separate file?

File: controllers/subscription.ts
Comment (ankushdharkar) [inline controllers/subscription.ts:50]: Better try-catch please, as mentioned above

File: routes/subscription.ts
Comment (ankushdharkar) [inline routes/subscription.ts:11]: What does this email route do?

**Review by ankushdharkar — CHANGES_REQUESTED** [review]

File: constants/subscription-validator.ts
Comment (tejaskh3) [reply constants/subscription-validator.ts:1]: This I  copied from internet I don't remember the site, may be it was my brave browser's auto generated regex or from Stackoverflow.

As we tested this is working fine for our usecase.

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:9]: Done

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:10]: Yes, it is same as we are validating this data at route level with the help of a middleware.

File: constants/subscription-validator.ts
Comment (tejaskh3) [reply constants/subscription-validator.ts:2]: update it, now we are taking any phone number and in the frontend also. 
And also as we discussed I have created a ticket to add country code flag later on.

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:34]: Done

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:49]: Yes this will be separate file, but for the time being we are only sending email to one person that's why I am using this here. Tough I'll add a TODO here.

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:50]: Done

File: routes/subscription.ts
Comment (tejaskh3) [reply routes/subscription.ts:11]: This is now changed to notify, used to send email notifications.

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:10]: Done

File: controllers/subscription.ts
Comment (iamitprakash) [inline controllers/subscription.ts:31]: 1. The dev check at the beginning of the function might be unnecessary or misplaced. If this is truly a development-only route, consider moving this check to the route definition rather than within the handler function.
2. Inconsistent error response format: In the if (!userAlreadySubscribed) check, the error response was using an object with a message property, while other error responses were using strings directly.
3. Inconsistent return in the catch block: The original code was using res.boom.badImplementation() without a return statement. This could lead to the function continuing execution after sending an error response.
4. Potential issue with error handling: The current error handling doesn't provide any specific information about the error. Depending on your requirements, you might want to log the error or provide more detailed error information (while being careful not to expose sensitive data).
5. Minor typo: "user unsubscribed successfully" has been capitalized to "User unsubscribed successfully" for consistency with other messages.

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: controllers/subscription.ts
Comment (tejaskh3) [reply controllers/subscription.ts:26]: 1. Added a middleware at route level for handling dev flag.
2. Now, we are only passing string in the message, as we only need the response.ok on the frontend. But for  **sendEmail** function it is returning message as well.
3. Added return statement to catch block as well.
4. Added logger while handling error.
5. Fixed this.

**Review by lakshayman — DISMISSED** [review]
LGTM

File: middlewares/devFlag.ts
Comment (iamitprakash) [inline middlewares/devFlag.ts:4]: 1. Dependency on boom: The code assumes you have a library named boom installed in your project with a notFound function that returns a 404 error.  If boom is not available, the catch block will be necessary to handle potential errors.

2. Error Handling:  While the current catch block passes the error (err) to the next function, it doesn't provide any specific handling within the middleware itself. You might want to consider logging the error or returning a more informative response depending on your application's needs.

3. Alternative for boom:  If you don't want to rely on boom, you can achieve the same functionality using vanilla Node.js functionalities:

File: middlewares/validators/subscription.ts
Comment (iamitprakash) [inline middlewares/validators/subscription.ts:6]: 1. Sanitization: Consider sanitizing the input data to prevent potential security vulnerabilities.
2. Asynchronous Validation: If you're dealing with asynchronous validation (e.g., checking if an email already exists in a database), you might need to use promises or async/await to handle the validation process correctly.

File: test/integration/subscription.test.js
Comment (iamitprakash) [inline test/integration/subscription.test.js:2]: -ve tests are missing

File: test/unit/middlewares/subscription-validator.test.js
Comment (iamitprakash) [inline test/unit/middlewares/subscription-validator.test.js:80]: what about should not return an error when phoneNumber is not into  correct format

**Review by iamitprakash — CHANGES_REQUESTED** [review]

File: middlewares/devFlag.ts
Comment (tejaskh3) [reply middlewares/devFlag.ts:4]: 1. Yes we are using **express-boom** to handle errors in the whole repo, this is why I  have used res.boom. I didn't use res.status just to maintain the consistency in the repo.
2. I have added a logger for handling errors with more information.
3.  Yes sir, we can either use ```res.status(404).send("Route not found")``` and many more libraries for this. But as in the whole repo we are using boom this is why I am also using this.

File: middlewares/validators/subscription.ts
Comment (tejaskh3) [reply middlewares/validators/subscription.ts:6]: 1. Now we are trimming the data and even checking them in regex as well.
2. In our current implementation, we handle the check for whether a user is already subscribed in the controller itself. This prevents users from subscribing again with the same email, ensuring proper validation without needing to handle it at the schema level right now.

File: test/unit/middlewares/subscription-validator.test.js
Comment (tejaskh3) [reply test/unit/middlewares/subscription-validator.test.js:80]: Sir, I have added test for sanitize as well as incorrect format phone number as well.

File: middlewares/validators/subscription.ts
Comment (vinit717) [inline middlewares/validators/subscription.ts:3]: Why comment?

File: constants/subscription-validator.ts
Comment (MehulKChaudhari) [reply constants/subscription-validator.ts:1]: @tejaskh3 You can use tools like this` https://regexr.com/` to build and test regex, before using it in code

Comment (tejaskh3) [reply constants/subscription-validator.ts:1]: Yes we both have tested this earlier.

**Review by iamitprakash — APPROVED** [review]

**Review by Achintya-Chatterjee — APPROVED** [review]

File: routes/subscription.ts
Comment (tejaskh3) [reply routes/subscription.ts]: Yes sir, it was decided.

File: config/custom-environment-variables.js
Comment (tejaskh3) [reply config/custom-environment-variables.js]: I didn't get your point, can you please explain me what you are talking.

File: middlewares/validators/subscription.ts
Comment (tejaskh3) [reply middlewares/validators/subscription.ts]: Thanks for pointing that out! While I understand that Joi provides built-in email validation, I’ve chosen to use a custom emailRegex here to have more  control over the format and edge cases we want to handle specifically.

**Review by iamitprakash — CHANGES_REQUESTED** [review]
please resolve all comments from Prakash sir

File: config/custom-environment-variables.js
Comment (tejaskh3) [reply config/custom-environment-variables.js]: Updated the name with correct formatting, please have a look sir.

**Review by Achintya-Chatterjee — APPROVED** [review]

File: config/custom-environment-variables.js
Comment (tejaskh3) [reply config/custom-environment-variables.js]: @prakashchoudhary07 sir said, you resolve this and I am good with it. I have changed the name so now we can resolve. Thanks sir

**Review by vinit717 — APPROVED** [review]

**Comment by tejaskh3** [comment]
successfully merged

**Pull request merged**
